README.md
# Calculator Component

This is a simple JavaScript calculator that is made to mimic the functionality
of the classic Windows calculator. It uses CSS classes for layout and unobtrusive
JavaScript for functionality. It does not rely on any external JavaScript libraries
to function.

See a live example at
[wjbryant.com/projects/javascript-calculator/](http://wjbryant.com/projects/javascript-calculator/ "JavaScript Calculator Demo").

## Features

* Support for input from both keyboard and mouse
* Keyboard shortcuts such as "Esc" for Clear and "r" for reciprocal
* Support for multiple calculators on a single page
* Public API for controlling calculators through JavaScript code

## How to Use the Calculator

First, download the zip file for the latest release and extract the contents into
a directory on your Web server. The two "min" files are compressed and should be
used in production. The other two js and css files are development versions which
are fully commented. The zip file also contains other documentation files. The
development and documentation files can be safely deleted if you do not need them.
The only two required files are the "min" files. They should be placed together
in the same directory.

To add the calculator to a web page, include the "min" js file in the page via
a script element. As of version 1.1a4, the script does not automatically create
calculators. Instead, call JSCALC.init() and any element with a class of "calc"
will have a calculator generated inside it.

Note that if you choose to rename or move the css file, it is necessary to set
the JSCALC.css property to the new file name (or full path if not in the same
directory) before calling JSCALC.init() so the script knows where to find it.
Check out the [documentation](http://wjbryant.github.com/JavaScript-Calculator/docs/ "API Documentation") and
[this post](http://wjbryant.com/2011/09/02/javascript-calculator-api/ "JavaScript Calculator API")
for information on more advanced usage.

### Example

```html
<div class="calc"></div>
<script src="calc-1.1-min.js"></script>
<script>JSCALC.init();</script>
```

## License

MIT

History.md
# Calculator Changelog

## Dev
* Make calculator into a component using https://github.com/component/component

## 1.1 (10/5/11)
* Minor documentation update

## 1.1b3 (10/1/11)
* Minor style improvements

## 1.1b2 (9/28/11)
* Moved CSS into a separate file
* Added the JSCALC.css property to allow overriding the default styles
* Fixed a bug that prevented calculators from being removed properly
* Various minor improvements

## 1.1b1 (9/26/11)
* Minor style improvements (button presses)
* Improved documentation

## 1.1a5 (9/9/11)
* Added the Calculator.focus() method
* Fixed a bug that prevented the current operation from being changed

## 1.1a4 (8/31/11)
* Added a public API
* Fixed the escape key shortcut so it works in all browsers
* Improved documentation
* Various minor improvements

## 1.1a3 (1/28/11)
* Initial public release

template.js
module.exports = '<div class="calcContainer">\n	<form action="">\n		<span class="calcTitle">Calculator</span>\n		<div>\n			<input type="text" class="calcDisplay" />\n			<input type="button" value="Backspace" class="calcClear calcFirst calcFunction" />\n			<input type="button" value="CE" class="calcClear calcFunction" />\n			<input type="button" value="C" class="calcClear calcFunction" />\n			<input type="button" value="7" class="calcFirst calcInput" />\n			<input type="button" value="8" class="calcInput" />\n			<input type="button" value="9" class="calcInput" />\n			<input type="button" value="/" class="calcFunction" />\n			<input type="button" value="sqrt" class="calcInput" />\n			<input type="button" value="4" class="calcFirst calcInput" />\n			<input type="button" value="5" class="calcInput" />\n			<input type="button" value="6" class="calcInput" />\n			<input type="button" value="*" class="calcFunction" />\n			<input type="button" value="%" class="calcInput" />\n			<input type="button" value="1" class="calcFirst calcInput" />\n			<input type="button" value="2" class="calcInput" />\n			<input type="button" value="3" class="calcInput" />\n			<input type="button" value="-" class="calcFunction" />\n			<input type="button" value="1/x" class="calcInput" />\n			<input type="button" value="0" class="calcFirst calcInput" />\n			<input type="button" value="+/-" class="calcInput" />\n			<input type="button" value="." class="calcInput" />\n			<input type="button" value="+" class="calcFunction" />\n			<input type="submit" value="=" class="calcFunction" />\n		</div>\n	</form>\n</div>';

calc-1.1-min.js
/*!
 * JavaScript Calculator v1.1
 * http://wjbryant.com/projects/javascript-calculator/
 *
 * Copyright (c) 2011 Bill Bryant
 * Licensed under the MIT license
 * http://opensource.org/licenses/mit-license.php
 */
var JSCALC;if(typeof JSCALC!=="object"){JSCALC={}}(function(){"use strict";var d=window,f=d.document,e="1.1",b={},a=d.onunload,h=false,c=0,g=f.getElementsByClassName?function(j){return f.getElementsByClassName(j)}:f.querySelectorAll?function(j){return f.querySelectorAll("."+j)}:f.evaluate?function(l){var j=f.evaluate("//*[contains(concat(' ', @class, ' '), ' "+l+" ')]",f,null,0,null),m=j.iterateNext(),k=[];while(m){k[k.length]=m;m=j.iterateNext()}return k}:function(n){var m,j=f.getElementsByTagName("*"),k=j.length,o=new RegExp("(^|\\s)"+n+"(\\s|$)"),l=[];for(m=0;m<k;m+=1){if(o.test(j[m].className)){l[l.length]=j[m]}}return l};function i(k){var j,n,t,u=0,p,y=true,l=/\./,s=null,r=false,w,x=false,v={},m=c;function q(z){switch(p){case"+":u+=z;break;case"-":u-=z;break;case"*":u*=z;break;case"/":u/=z;break}t.value=u}function o(D){D=D||d.event;var z,B,E,C,A=false;switch(D.type){case"keydown":w=B=D.keyCode;switch(B){case 27:z="C";break;case 8:z="Backspace";break;case 46:z="CE";break;default:return true}break;case"keypress":B=D.charCode||D.keyCode;if(B===13||B===9||w===37||w===39){return true}if(B===27||B===8||w===46){return false}if(w===188){z="."}else{z=String.fromCharCode(B).toLowerCase()}break;case"click":C=D.target||D.srcElement;if(C.tagName==="INPUT"&&C.type==="button"){z=C.value}else{return true}break;case"calculatorPressMethod":z=D.calckey;break;default:return true}E=parseFloat(t.value);switch(z){case"0":case"1":case"2":case"3":case"4":case"5":case"6":case"7":case"8":case"9":case".":if(!(z==="."&&l.test(t.value))){if(y){t.value="";y=false}t.value+=z}break;case"*":case"+":case"-":case"/":if(!x){if(u===0||s!==null){u=E}else{q(E)}s=null;r=true;y=true}p=z;A=true;break;case"C":t.blur();u=0;p="";y=true;s=null;r=false;t.value="0";break;case"CE":t.value="0";y=true;break;case"Backspace":t.value=t.value.slice(0,t.value.length-1);break;case"+/-":t.value=E*-1;break;case"%":if(E){t.value=u*E/100}break;case"sqrt":if(E>=0){t.value=Math.sqrt(E)}else{t.value="Invalid input for function"}break;case"a":case"c":case"v":case"x":if(D.ctrlKey){return true}break;case"1/x":case"r":if(E){t.value=1/E}else{t.value="Cannot divide by zero"}break;case"=":n.onsubmit();break}x=A;t.focus();return false}c+=1;k.innerHTML+='<div class="calcContainer"><form action=""><h1>Calculator</h1><div><input type="text" class="calcDisplay" /><input type="button" value="Backspace" class="calcClear calcFirst calcFunction" /><input type="button" value="CE" class="calcClear calcFunction" /><input type="button" value="C" class="calcClear calcFunction" /><input type="button" value="7" class="calcFirst calcInput" /><input type="button" value="8" class="calcInput" /><input type="button" value="9" class="calcInput" /><input type="button" value="/" class="calcFunction" /><input type="button" value="sqrt" class="calcInput" /><input type="button" value="4" class="calcFirst calcInput" /><input type="button" value="5" class="calcInput" /><input type="button" value="6" class="calcInput" /><input type="button" value="*" class="calcFunction" /><input type="button" value="%" class="calcInput" /><input type="button" value="1" class="calcFirst calcInput" /><input type="button" value="2" class="calcInput" /><input type="button" value="3" class="calcInput" /><input type="button" value="-" class="calcFunction" /><input type="button" value="1/x" class="calcInput" /><input type="button" value="0" class="calcFirst calcInput" /><input type="button" value="+/-" class="calcInput" /><input type="button" value="." class="calcInput" /><input type="button" value="+" class="calcFunction" /><input type="submit" value="=" class="calcFunction" /></div></form></div>';j=k.getElementsByTagName("form");n=j[j.length-1];t=n.getElementsByTagName("input")[0];t.setAttribute("autocomplete","off");t.value="0";t.onkeydown=t.onkeypress=n.onclick=o;n.onsubmit=function(){if(r){s=parseFloat(t.value)||0;r=false}q(s);y=true;t.focus();return false};v.focus=function(){t.focus()};v.press=function(B){var C,z,A;if(typeof B==="number"){C=B.toString().split("")}else{if(typeof B==="string"&&B){C=[B]}else{return this}}z=C.length;for(A=0;A<z;A+=1){o({type:"calculatorPressMethod",calckey:C[A]})}return this};v.remove=function(){t.onkeydown=t.onkeypress=n.onclick=null;k.removeChild(n.parentNode);delete b[m];v=null};v.container=k;b[m]=v;return v}JSCALC.get=function(j){if(!j||j.nodeType!==1){return}var m,l=b,k;for(m in l){if(l.hasOwnProperty(m)){if(j===l[m].container){k=l[m];break}}}return k||false};JSCALC.getCalcs=function(){var l,k=[],j=b;for(l in j){if(j.hasOwnProperty(l)){k[k.length]=j[l]}}return k};JSCALC.init=function(l){var m,s=[],u=false,k,t,r,o,n,j,q,p=[];if(typeof l==="string"){l=f.getElementById(l)}if(typeof l==="object"&&l.nodeType===1){if(l.className){if(l.className.indexOf("calc")===-1){l.className+=" calc"}}else{l.className="calc"}s[0]=l;u=true}else{s=g("calc")}t=s.length;if(t){if(!h){if(!JSCALC.css){o=f.getElementsByTagName("script");n=o.length;for(r=0;r<n;r+=1){j=o[r].src;if(j.indexOf("calc-"+e)!==-1){q=j.lastIndexOf("/");JSCALC.css=(q!==-1?j.slice(0,q+1):"")+"calc-"+e+"-min.css";break}}}if(JSCALC.css){m=f.createElement("link");m.type="text/css";m.rel="stylesheet";m.href=JSCALC.css;(f.head||f.getElementsByTagName("head")[0]).appendChild(m);h=true}}for(r=0;r<t;r+=1){k=s[r];if(!JSCALC.get(k)){p[p.length]=i(k)}}}return u?(p[0]||false):p};JSCALC.removeAll=function(){var k,j=b;for(k in j){if(j.hasOwnProperty(k)){j[k].remove()}}};if(d.attachEvent){d.attachEvent("onunload",JSCALC.removeAll)}else{d.onunload=function(j){JSCALC.removeAll();if(a){return a(j)}}}}());

src/calc-1.1.js
/*!
 * JavaScript Calculator v1.1
 * http://wjbryant.com/projects/javascript-calculator/
 *
 * Copyright (c) 2011 Bill Bryant
 * Licensed under the MIT license
 * http://opensource.org/licenses/mit-license.php
 */

/**
 * @fileOverview JavaScript Calculator
 * @author       Bill Bryant
 * @version      1.1
 */

/*jslint browser: true, maxerr: 50, indent: 4 */

 /**
  * The JSCALC "namespace" / global object.
  *
  * @namespace
  */
var JSCALC;
if (typeof JSCALC !== 'object') {
    JSCALC = {};
}

(function () {
    'use strict';

    var win = window,
        document = win.document,
        version = '1.1',
        calculators = {}, // an object containing all the calculators created
        origUnload = win.onunload,
        styleAdded = false,
        nextID = 0,

        /**
         * A simple implementation of getElementsByClassName. Checks for one
         * class name and does not account for differences between NodeList and
         * Array.
         *
         * @param  {String}         className the class name
         * @return {NodeList|Array}           a NodeList or Array of elements
         *
         * @function
         * @ignore
         */
        getElementsByClassName = document.getElementsByClassName ?
                function (className) {
                    return document.getElementsByClassName(className);
                } :
                document.querySelectorAll ?
                        function (className) {
                            return document.querySelectorAll('.' + className);
                        } :
                        document.evaluate ?
                                function (className) {
                                    var elems = document.evaluate(
                                            "//*[contains(concat(' ', @class, ' '), ' " +
                                                className + " ')]",
                                            document,
                                            null,
                                            0,
                                            null
                                        ),
                                        elem = elems.iterateNext(),
                                        results = [];
                                    while (elem) {
                                        results[results.length] = elem;
                                        elem = elems.iterateNext();
                                    }
                                    return results;
                                } :
                                function (className) {
                                    var i,
                                        elems = document.getElementsByTagName('*'),
                                        num = elems.length,
                                        pattern = new RegExp('(^|\\s)' + className + '(\\s|$)'),
                                        results = [];
                                    for (i = 0; i < num; i += 1) {
                                        if (pattern.test(elems[i].className)) {
                                            results[results.length] = elems[i];
                                        }
                                    }
                                    return results;
                                };

    /**
     * Creates a new calculator in the specified container element (module).
     *
     * @param  {Element} calcMod the element to contain the calculator
     * @return {Object}          a Calculator object
     *
     * @ignore
     */
    function createCalc(calcMod) {
        var forms,
            form,
            display,
            total = 0,
            operation,
            clearNext = true,
            dot = /\./,
            lastNum = null,
            getLastNum = false,
            lastKeyDown,
            operationPressed = false, // whether an operation was the last key pressed
            calcObj = {},
            id = nextID;

        /**
         * Performs the basic mathematical operations (addition, subtraction,
         * multiplication, division) on the current total with the given
         * value for the current operation.
         *
         * @param {Number} val the value to use in the calculation with the total
         *
         * @ignore
         */
        function calc(val) {
            switch (operation) {
            case '+':
                total += val;
                break;
            case '-':
                total -= val;
                break;
            case '*':
                total *= val;
                break;
            case '/':
                total /= val;
                break;
            }
            display.value = total;
        }

        /**
         * This function handles input for the form's keydown, keypress and
         * click events. Any keypresses that are not related to a calculator
         * function are not allowed.
         *
         * @return {Boolean} whether the default action is allowed
         *
         * @ignore
         */
        function handleInput(e) {
            e = e || win.event;

            var key, // the key (char) that was pressed / clicked
                code, // the key code
                val, // the numeric value of the display
                target, // the target element of the event
                isOperation = false; // whether the button pressed is an operation

            // this switch statement sets the key variable
            switch (e.type) {
            case 'keydown':
                lastKeyDown = code = e.keyCode;

                switch (code) {
                case 27:
                    // escape
                    key = 'C';
                    break;
                case 8:
                    // backspace
                    key = 'Backspace';
                    break;
                case 46:
                    // delete
                    key = 'CE';
                    break;
                default:
                    // allow all other keys (enter, tab, numbers, letters, etc.)
                    return true;
                }
                break;
            case 'keypress':
                // most browsers provide the charCode property when the keypress event is fired
                // IE and Opera provide the character code in the keyCode property instead
                code = e.charCode || e.keyCode;

                // this event is fired for all keys in Firefox and Opera
                // because of this, we need to check the last keyCode
                // from the keydown event for certain keys

                // allow enter, tab and left and right arrow keys
                // the enter key fires the keypress event in all browsers
                // the other keys are handled here for Firefox and Opera
                if (code === 13 || code === 9 || lastKeyDown === 37 || lastKeyDown === 39) {
                    return true;
                }
                // these keys are handled on keydown (escape, backspace, delete)
                // this is for Firefox and Opera (and sometimes IE for the escape key)
                if (code === 27 || code === 8 || lastKeyDown === 46) {
                    return false;
                }

                // get the key character in lower case
                if (lastKeyDown === 188) {
                    key = '.'; // allow the comma key to be used for a decimal point
                } else {
                    key = String.fromCharCode(code).toLowerCase();
                }
                break;
            case 'click':
                target = e.target || e.srcElement;
                if (target.tagName === 'INPUT' && target.type === 'button') {
                    key = target.value;
                } else {
                    return true;
                }
                break;
            case 'calculatorPressMethod':
                // special case for the press method of the calculator object
                key = e.calckey;
                break;
            default:
                // the switch statement should never make it here
                // this is just a safeguard
                return true;
            }

            val = parseFloat(display.value);

            switch (key) {
            case '0':
            case '1':
            case '2':
            case '3':
            case '4':
            case '5':
            case '6':
            case '7':
            case '8':
            case '9':
            case '.':
                // don't allow more than one decimal point
                if (!(key === '.' && dot.test(display.value))) {
                    if (clearNext) {
                        display.value = '';
                        clearNext = false;
                    }
                    display.value += key;
                }
                break;
            case '*':
            case '+':
            case '-':
            case '/':
                // if an operation was the last key pressed,
                // do nothing but change the current operation
                if (!operationPressed) {
                    if (total === 0 || lastNum !== null) {
                        total = val;
                    } else {
                        calc(val);
                    }
                    lastNum = null;
                    getLastNum = true;
                    clearNext = true;
                }
                operation = key;
                isOperation = true;
                break;
            case 'C':
                display.blur(); // so Firefox can clear display when escape key is pressed
                total = 0;
                operation = '';
                clearNext = true;
                lastNum = null;
                getLastNum = false;
                display.value = '0';
                break;
            case 'CE':
                display.value = '0';
                clearNext = true;
                break;
            case 'Backspace':
                display.value = display.value.slice(0, display.value.length - 1);
                break;
            case '+/-':
                display.value = val * -1;
                break;
            case '%':
                if (val) {
                    display.value = total * val / 100;
                }
                break;
            case 'sqrt':
                if (val >= 0) {
                    display.value = Math.sqrt(val);
                } else {
                    display.value = 'Invalid input for function';
                }
                break;
            case 'a':
            case 'c':
            case 'v':
            case 'x':
                // allow select all, copy, paste and cut key combinations
                if (e.ctrlKey) {
                    return true;
                }
                break;
            case '1/x':
            case 'r':
                if (val) {
                    display.value = 1 / val;
                } else {
                    display.value = 'Cannot divide by zero';
                }
                break;
            case '=':
                form.onsubmit();
                break;
            }

            operationPressed = isOperation;
            display.focus();
            return false;
        }

        // increment the ID counter
        nextID += 1;

        // create the calculator elements
        calcMod.innerHTML += '<div class="calcContainer"><form action=""><h1>Calculator</h1><div><input type="text" class="calcDisplay" /><input type="button" value="Backspace" class="calcClear calcFirst calcFunction" /><input type="button" value="CE" class="calcClear calcFunction" /><input type="button" value="C" class="calcClear calcFunction" /><input type="button" value="7" class="calcFirst calcInput" /><input type="button" value="8" class="calcInput" /><input type="button" value="9" class="calcInput" /><input type="button" value="/" class="calcFunction" /><input type="button" value="sqrt" class="calcInput" /><input type="button" value="4" class="calcFirst calcInput" /><input type="button" value="5" class="calcInput" /><input type="button" value="6" class="calcInput" /><input type="button" value="*" class="calcFunction" /><input type="button" value="%" class="calcInput" /><input type="button" value="1" class="calcFirst calcInput" /><input type="button" value="2" class="calcInput" /><input type="button" value="3" class="calcInput" /><input type="button" value="-" class="calcFunction" /><input type="button" value="1/x" class="calcInput" /><input type="button" value="0" class="calcFirst calcInput" /><input type="button" value="+/-" class="calcInput" /><input type="button" value="." class="calcInput" /><input type="button" value="+" class="calcFunction" /><input type="submit" value="=" class="calcFunction" /></div></form></div>';

        // get the calculator inputs
        forms = calcMod.getElementsByTagName('form');
        form = forms[forms.length - 1]; // make sure it's the one that was just added
        display = form.getElementsByTagName('input')[0];
        display.setAttribute('autocomplete', 'off');
        display.value = '0';
        display.onkeydown = display.onkeypress = form.onclick = handleInput;

        /**
         * Calculates the value of the last entered operation and displays the result.
         *
         * @return {Boolean} always returns false to prevent the form from being submitted
         *
         * @ignore
         */
        form.onsubmit = function () {
            if (getLastNum) {
                lastNum = parseFloat(display.value) || 0;
                getLastNum = false;
            }
            calc(lastNum);
            clearNext = true;
            display.focus();
            return false;
        };

        /**
         * The Calculator class. Calculator objects are returned by some
         * methods of the JSCALC object.
         *
         * @name Calculator
         * @class
         * @private
         */

        /**
         * Gives focus to the calculator display.
         *
         * @function
         * @name focus
         * @memberOf Calculator.prototype
         */
        calcObj.focus = function () {
            display.focus();
        };

        /**
         * Simulates pressing a button on the calculator.
         *
         * @param  {Number|String} button the button(s) to press - a number can
         *                                represent multiple buttons, but a
         *                                string can only represent one
         * @return {Object}               the Calculator object
         *
         * @function
         * @name press
         * @memberOf Calculator.prototype
         */
        calcObj.press = function (button) {
            var buttons,
                num,
                i;

            // if button is a number, convert it to an array of digits as strings
            if (typeof button === 'number') {
                buttons = button.toString().split('');
            } else if (typeof button === 'string' && button) {
                buttons = [button];
            } else {
                // invalid argument
                return this; // do nothing, but still allow method chaining
            }

            num = buttons.length;
            for (i = 0; i < num; i += 1) {
                handleInput({
                    type: 'calculatorPressMethod',
                    calckey: buttons[i]
                });
            }

            return this; // allow method chaining
        };

        /**
         * Removes the calculator and sets the Calculator object to null.
         *
         * @function
         * @name remove
         * @memberOf Calculator.prototype
         */
        calcObj.remove = function () {
            display.onkeydown = display.onkeypress = form.onclick = null;
            calcMod.removeChild(form.parentNode);
            delete calculators[id];
            calcObj = null;
        };

        /**
         * A reference to the element that contains the calculator.
         *
         * @name container
         * @memberOf Calculator.prototype
         */
        calcObj.container = calcMod;

        calculators[id] = calcObj; // keep track of all calculators

        return calcObj;
    }

    /**
     * Gets the Calculator object associated with the calculator contained in
     * the specified element.
     *
     * @param  {Element} container the element containing the calculator
     * @return {Object}            the Calculator object or false if none exists
     */
    JSCALC.get = function (container) {
        // if the container argument is not an element node, do nothing
        if (!container || container.nodeType !== 1) {
            return;
        }

        var id,
            calcs = calculators,
            calc;

        for (id in calcs) {
            if (calcs.hasOwnProperty(id)) {
                if (container === calcs[id].container) {
                    calc = calcs[id];
                    break;
                }
            }
        }

        return calc || false;
    };

    /**
     * Gets the Calculator objects for all the calculators on the page.
     *
     * @return {Array} an array of Calculator objects
     */
    JSCALC.getCalcs = function () {
        var id,
            calcArray = [],
            calcs = calculators;

        // the calculators array may be sparse, so copy all objects from it
        // into a dense array and return that instead
        for (id in calcs) {
            if (calcs.hasOwnProperty(id)) {
                calcArray[calcArray.length] = calcs[id];
            }
        }

        return calcArray;
    };

    /**
     * Specifies the CSS file location. By default, this property is set
     * automatically the first time JSCALC.init() is called. Set this property
     * before calling JSCALC.init() to override the automatic value.
     *
     * @name css
     * @memberOf JSCALC
     */

    /**
     * Creates calculators.
     *
     * @param  {String|Element} elem (optional) the element in which to create the calculator
     * @return {Object|Array}        If an argument is specified, the Calculator object or
     *                               false is returned. If no (or invalid) arguments are
     *                               specified, an array of Calculator objects or an empty
     *                               array is returned.
     */
    JSCALC.init = function (elem) {
        var calcStyle,
            calcMods = [],
            args = false,
            calcMod,
            len,
            i,
            scripts,
            numScripts,
            src,
            separatorPos,
            newCalcs = [];

        // treat a string argument as an element id
        if (typeof elem === 'string') {
            elem = document.getElementById(elem);
        }

        // if the argument is an element object or an element was found by id
        if (typeof elem === 'object' && elem.nodeType === 1) {
            // add the "calc" class name to the element
            if (elem.className) {
                if (elem.className.indexOf('calc') === -1) {
                    elem.className += ' calc';
                }
            } else {
                elem.className = 'calc';
            }

            // add the element to the array of calculator modules to be initialized
            calcMods[0] = elem;
            args = true;
        } else {
            // if an element node was not found or specified, get all elements
            // with a class name of "calc"
            calcMods = getElementsByClassName('calc');
        }

        len = calcMods.length;

        // if there is at least one element in the array
        if (len) {
            // add the style sheet if it isn't already added
            if (!styleAdded) {
                // try to guess the CSS file location if it is not specified
                if (!JSCALC.css) {
                    scripts = document.getElementsByTagName('script');
                    numScripts = scripts.length;

                    // loop through the script elements and find this script by
                    // checking the src property
                    for (i = 0; i < numScripts; i += 1) {
                        src = scripts[i].src;

                        if (src.indexOf('calc-' + version) !== -1) {
                            // guess the CSS file path based on the js file path
                            separatorPos = src.lastIndexOf('/');
                            JSCALC.css = (separatorPos !== -1 ? src.slice(0, separatorPos + 1) : '') +
                                'calc-' + version + '-min.css';
                            break;
                        }
                    }
                }

                // make sure the stylesheet location is set before adding the stylesheet
                if (JSCALC.css) {
                    calcStyle = document.createElement('link');
                    calcStyle.type = 'text/css';
                    calcStyle.rel = 'stylesheet';
                    calcStyle.href = JSCALC.css;
                    (document.head || document.getElementsByTagName('head')[0]).appendChild(calcStyle);
                    styleAdded = true;
                }
            }

            // loop through the array and create a calculator in each element
            for (i = 0; i < len; i += 1) {
                calcMod = calcMods[i];

                // check to ensure a calculator does not already exist in the
                // specified element
                if (!JSCALC.get(calcMod)) {
                    newCalcs[newCalcs.length] = createCalc(calcMod);
                }
            }
        }

        // if an argument was specified, return a single object or false if one
        // could not be created
        // else, return the array of objects even if it is empty
        return args ? (newCalcs[0] || false) : newCalcs;
    };

    /**
     * Removes all calculators.
     */
    JSCALC.removeAll = function () {
        var id,
            calcs = calculators;

        // remove each calculator in the calculators array (calcs)
        for (id in calcs) {
            if (calcs.hasOwnProperty(id)) {
                calcs[id].remove();
            }
        }
    };

    // remove all event handlers on window unload
    // this will prevent a memory leak in older versions of IE
    if (win.attachEvent) {
        win.attachEvent('onunload', JSCALC.removeAll);
    } else {
        win.onunload = function (e) {
            JSCALC.removeAll();
            // preserve original window unload event handler
            if (origUnload) {
                return origUnload(e);
            }
        };
    }
}());